mpcs/src/basefold/commit_phase.rs

```diff
@@ -201,54 +325,142 @@ where
+        .unwrap_or_else(|| {
+            codewords
+                .iter()
+                .map(|v| v.values.len())
+                .max()
+                .expect("empty codeword")
+        });
+    let next_level_target_len = target_len >> 1;
+    let level = log2_strict_usize(target_len) - 1;
     let folding_coeffs =
         <Spec::EncodingScheme as EncodingScheme<E>>::prover_folding_coeffs_level(pp, level);
-    debug_assert_eq!(folding_coeffs.len(), 1 << level);
     let inv_2 = E::BaseField::from_u64(2).inverse();
-    RowMajorMatrix::new(
-        values
-            .par_chunks_exact(2)
-            .zip(folding_coeffs)
-            .map(|(ys, coeff)| {
-                let (left, right) = (ys[0], ys[1]);
-                // original (left, right) = (lo + hi*x, lo - hi*x), lo, hi are codeword, but after times x it's not codeword
-                // recover left & right codeword via (lo, hi) = ((left + right) / 2, (left - right) / 2x)
-                let (lo, hi) = ((left + right) * inv_2, (left - right) * *coeff); // e.g. coeff = (2 * dit_butterfly)^(-1) in rs code
-                // we do fold on folded = (1-r) * left_codeword + r * right_codeword, as it match perfectly with raw message in lagrange domain fixed variable
-                lo + (hi - lo) * challenge
-            })
-            .collect::<Vec<_>>(),
-        2,
-    )
+    debug_assert_eq!(folding_coeffs.len(), 1 << level);
+
+    // take codewrods match with target length then fold
+    let codewords_matched =
+        pop_front_while(codewords, |codeword| codeword.values.len() == target_len);
+    // take codewrods match next target length in preparation of being committed together
+    let codewords_next_level_matched = pop_front_while(codewords, |codeword| {
+        codeword.values.len() == next_level_target_len
+    });
+
+    // optimize for single codeword match
+    let folded_codeword = if (running_codeword_opt.is_some() as usize + codewords_matched.len())
+        == 1
+        && codewords_next_level_matched.is_empty()
+    {
+        RowMajorMatrix::new(
+            running_codeword_opt
+                .or_else(|| codewords_matched.first().map(|m| m.as_view()))
+                .unwrap()
+                .values
+                .par_chunks_exact(2)
+                .zip(folding_coeffs)
+                .map(|(ys, coeff)| {
+                    let (left, right) = (ys[0], ys[1]);
+                    // original (left, right) = (lo + hi*x, lo - hi*x), lo, hi are codeword, but after times x it's not codeword
+                    // recover left & right codeword via (lo, hi) = ((left + right) / 2, (left - right) / 2x)
+                    let (lo, hi) = ((left + right) * inv_2, (left - right) * *coeff); // e.g. coeff = (2 * dit_butterfly)^(-1) in rs code
+                    // we do fold on folded = (1-r) * left_codeword + r * right_codeword, as it match perfectly with raw message in lagrange domain fixed variable
```

Comment:
nitpick: should be "we do fold on (lo, hi) to get folded = (1-r) * lo + r* hi (with lo, hi are two codewords) ...."